chalk-ir/src/lib.rs

```diff
@@ -2898,3 +2939,16 @@ pub struct AnswerSubst<I: Interner> {
     /// Delayed subgoals, used when the solver answered with an (incomplete) `Answer` (instead of a `CompleteAnswer`).
     pub delayed_subgoals: Vec<InEnvironment<Goal<I>>>,
 }
+
+/// Logic to decide the Variance for a given subst
+pub trait UnificationDatabase<I>
+where
+    Self: std::fmt::Debug,
+    I: Interner,
+{
+    /// Gets the variances for the substitution of a fn def
+    fn fn_def_variance(&self, fn_def_id: FnDefId<I>) -> Vec<Variance>;
```

Comment:
`I::Variances` would probably be better

Reply:
Hmm true. But, I honestly would prefer to avoid creating these at all. I'm wondering if maybe a
```
fn with_fn_def_variances<R>(&self, fn_def_id: FnDefId<I>, op: impl Fn(&[Variance])) -> R;
```
would be better.

Reply:
That is not possible with a `dyn` trait. But there could be something like `&self -> &[Variance]`, potentially, though that's kind of limiting.

Reply:
Well, we actually have to get this via `RustIrDatabase::unification_database()`, so it doesn't have to be `dyn`.

Reply:
Well, we're not generic over the `RustIrDatabase`, so that means it must be `dyn`. Anyway I think we *should* make it `dyn` or at least `dyn` compatible.